src/structs/common_macros.rs

```diff
@@ -79,156 +129,220 @@ macro_rules! pointwise_sub(
                 $( self.$compN -= right.$compN; )+
             }
         }
+
+        ref_binop!(impl Sub, sub for $t; $($compN),+);
     )
 );
 
 
-macro_rules! pointwise_scalar_mul(
-    ($t: ident, $($compN: ident),+) => (
-        impl<N: Copy + Mul<N, Output = N>> Mul<N> for $t<N> {
-            type Output = $t<N>;
+macro_rules! ref_binop_scalar_exact(
+    (impl $imp:ident<$lhs:ident>, $method:ident for $t:ident; $($compN:ident),+) => {
+        impl $imp<$t<$lhs>> for $lhs {
+            type Output = $t<$lhs>;
 
             #[inline]
-            fn mul(self, right: N) -> $t<N> {
-                $t::new($(self.$compN * right),+)
+            fn $method(self, right: $t<$lhs>) -> $t<$lhs> {
+                $t::new(
+                    $(
+                        $imp::$method(right.$compN, self)
+                    ),+
+                )
             }
         }
 
-        impl<N: Copy + MulAssign<N>> MulAssign<N> for $t<N> {
+        impl<'a, 'b> $imp<&'a $t<$lhs>> for &'b $lhs {
+            type Output = $t<$lhs>;
+
             #[inline]
-            fn mul_assign(&mut self, right: N) {
-                $( self.$compN *= right; )+
+            fn $method(self, right: &'a $t<$lhs>) -> $t<$lhs> {
+                $t::new(
+                    $(
+                        $imp::$method(right.$compN, self)
+                    ),+
+                )
             }
         }
 
-        impl Mul<$t<f32>> for f32 {
-            type Output = $t<f32>;
+        impl<'a> $imp<$t<$lhs>> for &'a $lhs {
+            type Output = $t<$lhs>;
 
             #[inline]
-            fn mul(self, right: $t<f32>) -> $t<f32> {
-                $t::new($(self * right.$compN),+)
+            fn $method(self, right: $t<$lhs>) -> $t<$lhs> {
+                $t::new(
+                    $(
+                        $imp::$method(right.$compN, self)
+                    ),+
+                )
             }
         }
 
-        impl Mul<$t<f64>> for f64 {
-            type Output = $t<f64>;
+        impl<'a> $imp<&'a $t<$lhs>> for $lhs {
+            type Output = $t<$lhs>;
 
             #[inline]
-            fn mul(self, right: $t<f64>) -> $t<f64> {
-                $t::new($(self * right.$compN),+)
+            fn $method(self, right: &'a $t<$lhs>) -> $t<$lhs> {
+                $t::new(
+                    $(
+                        $imp::$method(right.$compN, self)
+                    ),+
+                )
             }
         }
-    )
+    };
 );
 
 
-macro_rules! pointwise_scalar_div(
-    ($t: ident, $($compN: ident),+) => (
-        impl<N: Copy + Div<N, Output = N>> Div<N> for $t<N> {
+macro_rules! ref_binop_scalar(
+    // Symmetric to `std::ops::forward_ref_binop!`
+    (impl $imp:ident, $method:ident for $t:ident; $($compN:ident),+) => {
+        ref_binop_scalar_exact!(impl $imp<f32>, $method for $t; $($compN),+);
+        ref_binop_scalar_exact!(impl $imp<f64>, $method for $t; $($compN),+);
```

Comment:
@sebcrozet I re-implemented the `f32 * T<f32>` ops, now it should support `&f32 * T<f32>`, `f32 * &T<f32>` and `&f32 * &T<f32>` aswell.